lib-src/libcs/getdouble.3
.\" COPYRIGHT NOTICE
.\" Copyright (c) 1994 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" See <cmu_copyright.h> for use and distribution information.
.\" 
.\" 
.\" HISTORY
.\" getdouble.3,v
.\" Revision 1.1.1.1  1996/11/22 19:19:44  rvb
.\"  almost done
.\"
.\" Revision 1.1  1996/11/22 19:19:44  braam
.\" First Checkin (pre-release)
.\"
Revision 1.1  96/06/03  19:00:27  satya
Initial revision

.\" Revision 1.2  1995/10/18  14:58:21  moore
.\" 	Created libcs man(3) pages from mach entries
.\" 	[1995/10/18  14:54:09  moore]
.\"
.\" $EndLog$
.\" Copyright (c) 1990 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" Permission to use, copy, modify and distribute this software and its
.\" documentation is hereby granted, provided that both the copyright
.\" notice and this permission notice appear in all copies of the
.\" software, derivative works or modified versions, and any portions
.\" thereof, and that both notices appear in supporting documentation.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
.\" DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
.\" SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
.\" INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
.\" RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
.\" CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
.\" CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" Users of this software agree to return to Carnegie Mellon any
.\" improvements or extensions that they make and grant Carnegie the
.\" rights to redistribute these changes.
.\"
.\" Export of this software is permitted only after complying with the
.\" regulations of the U.S. Deptartment of Commerce relating to the
.\" Export of Technical Data.
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" HISTORY
.\" getdouble.3,v
.\" Revision 1.1.1.1  1996/11/22 19:19:44  rvb
.\"  almost done
.\"
.\" Revision 1.1  1996/11/22 19:19:44  braam
.\" First Checkin (pre-release)
.\"
Revision 1.1  96/06/03  19:00:27  satya
Initial revision

.\" Revision 1.2  1995/10/18  14:58:21  moore
.\" 	Created libcs man(3) pages from mach entries
.\" 	[1995/10/18  14:54:09  moore]
.\"
.\" Revision 1.1.1.2  1995/10/18  14:54:09  moore
.\" 	Created libcs man(3) pages from mach entries
.\"
.\" Revision 1.2  90/12/12  15:50:44  mja
.\" 	Add copyright/disclaimer for distribution.
.\" 
.\" 13-Nov-86  Andi Swimmer (andi) at Carnegie-Mellon University
.\" 	Revised for 4.3.
.\" 
.\" 23-Oct-82  Steven Shafer (sas) at Carnegie-Mellon University
.\" 	Error or EOF on standard input now causes default value to be used.
.\" 
.\" 05-Dec-79  Steven Shafer (sas) at Carnegie-Mellon University
.\" 	Created.
.\" 
.TH GETDOUBLE 3 10/23/82
.CM 3
.SH "NAME"
getdouble, getfloat \- ask user to type floating-point value
.SH "SYNOPSIS"
.B
#include <ctype.h>
.br
.B
#include <math.h>
.br
.B
#include <stdio.h>
.sp
.B
double getdouble (prompt,min,max,defalt);
.br
.B
char *prompt;
.br
.B
double min,max,defalt;
.sp
.B
float getfloat (prompt,min,max,defalt);
.br
.B
char *prompt;
.br
.B
float min,max,defalt;
.SH "DESCRIPTION"
.I
Getdouble
and
.I
getfloat
ask the user to type a floating-point number.
They begin
by printing the string
.I
prompt
as a message to the user.
The user then types in a floating-point
number which is parsed.
If the number is valid and is within
the range
.I
min
to
.I
max,
inclusive, then the value is returned as the value of
.I
getdouble
or
.I
getfloat.
If the value is invalid or out of range, then an error message is
printed and the cycle is repeated.
If the user types just a carriage
return, then the value
.I
defalt
is returned.
.sp
.I
Getdouble
and
.I
getfloat
are identical, except for the type of the parameters and the results.
.SH "LIBRARY NOTES"
You must link against /usr/local/lib/libcs.a to define this routine.
.SH "SEE ALSO"
.IR getint (3), 
.IR getbool (3), 
etc.
.br
.IR doublearg( 3), 
.IR floatarg (3)


NEWS
Incompatible RVM update (Coda client)
-------------------------------------

2008/02/06 RVM version 39
    Move store identifier from per-volume to global rvg struct.

2005/06/07 RVM version 38
    Added realm::realmid (as we can't use &realm_object on 64-bit systems)

2005/06/16 RVM version 37
    Removed fsdb::device and fsobj::inode


lib-src/libcs/getdouble.c
#ifndef _BLURB_
#define _BLURB_
/*
 * This code was originally part of the CMU SCS library "libcs".
 * A branch of that code was taken in May 1996, to produce this
 * standalone version of the library for use in Coda and Odyssey
 * distribution.  The copyright and terms of distribution are
 * unchanged, as reproduced below.
 *
 * Copyright (c) 1990-96 Carnegie Mellon University
 * All Rights Reserved.
 * 
 * Permission to use, copy, modify and distribute this software and its
 * documentation is hereby granted, provided that both the copyright
 * notice and this permission notice appear in all copies of the
 * software, derivative works or modified versions, and any portions
 * thereof, and that both notices appear in supporting documentation.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
 * SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
 * INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
 * RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
 * CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Users of this software agree to return to Carnegie Mellon any
 * improvements or extensions that they make and grant Carnegie the
 * rights to redistribute these changes.
 *
 * Export of this software is permitted only after complying with the
 * regulations of the U.S. Deptartment of Commerce relating to the
 * Export of Technical Data.
 */

static char *rcsid = "$Header: /afs/cs.cmu.edu/project/coda-braam/src/coda-4.0.1/RCSLINK/./lib-src/libcs/getdouble.c,v 1.1 1996/11/22 19:19:27 braam Exp $";
#endif /*_BLURB_*/

/*  getdouble --  prompt user for double
 * 
 *  Usage:  f = getdouble (prompt,min,max,defalt)
 * 	double f,min,max,defalt;
 * 	char *prompt;
 * 
 *  Getdouble prints the message:  prompt  (min to max)  [defalt]
 *  and accepts a line of input from the user.  If the input
 *  is not null or numeric, an error message is printed; otherwise,
 *  the value is converted to a double (or the value "defalt" is
 *  substituted if the input is null).  Then, the value is
 *  checked to ensure that is lies within the range "min" to "max".
 *  If it does not, an error message is printed.  As long as
 *  errors occur, the cycle is repeated; when a legal value is
 *  entered, this value is returned by getdouble.
 *  On end-of-file or read error in the standard input, the default
 *  is returned.
 * 
 */

#include <stdio.h>
#include <ctype.h>
#include <math.h>

#include "libcs.h"

double getdouble (prompt,min,max,defalt)
	double min,max,defalt;
const	char *prompt;
{
	char input [200];
	register char *p;
	register int err;
	double i = 0.0;

	fflush (stdout);
	do {

		fprintf (stderr,"%s  (%g to %g)  [%g]  ",prompt,min,max,defalt);
		fflush (stderr);

		err = 0;
		if (gets(input) == NULL) {
			i = defalt;
			err = (i < min || max < i);
		}
		else {
			for (p=input; *p &&
		 	(isdigit(*p) || *p=='-' || *p=='.' || *p=='+'
		  	|| *p=='e' || *p=='E');
		 	p++);
	
			if (*p) {		/* non-numeric */
				err = 1;
			} else {
				if (*input)	i = atof (input);
				else		i = defalt;
				err = (i < min || max < i);
			}
		}

		if (err) fprintf (stderr,"Must be a number between %g and %g\n",
			min,max);
	} while (err);

	return (i);
}
